Remove old model usage from MailPoet\WooCommerce\Subscription and WooCommerceBlocksIntegration [MAILPOET-4365]

rodrigoprimo opened the pull request:
## Description
N/A

## QA notes
Make sure that users can still subscribe to lists when using the WooCommerce and WooCommerce Blocks checkouts.

## Linked tickets
[MAILPOET-4365]

[MAILPOET-4365]: https://mailpoet.atlassian.net/browse/MAILPOET-4365?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

rodrigoprimo committed:
Replace Paris models with Doctrine entities in WooCommerceBlocksIntegration

[MAILPOET-4365]

rodrigoprimo committed:
Use SubscriberEntity instead of the Subscriber model to get statuses

[MAILPOET-4365]

rodrigoprimo committed:
Add method to get segments that a subscriber is subscribed to

[MAILPOET-4365]

rodrigoprimo committed:
Use getServiceWithOverrides() to instantiated tested class

Doing this instead of calling the constructor of the class directly to
avoid having to update the test class whenever there is a change to the
constructor of the tested class.

[MAILPOET-4365]

rodrigoprimo committed:
Replace Paris models with Doctrine entities in Subscription::subscribeOnCheckout()

Besides Subscription::subscribeOnCheckout(), this commit also replaces
Paris models with Doctrine entities in all the methods that are called
by Subscription::subscribeOnCheckout().

[MAILPOET-4365]

rodrigoprimo committed:
Replace remaining uses of the Segment model with Doctrine in Subscription

[MAILPOET-4365]

rodrigoprimo committed:
Replace SubscriberSegment model with Doctrine code in Subscription

[MAILPOET-4365]

rodrigoprimo committed:
Make sure the subscriber_segment table is truncated

This is necessary while
https://mailpoet.atlassian.net/browse/MAILPOET-4580. Without this
change, leaving entries in wp_mailpoet_subscriber_segment has the
potential to interfere with other tests as it happened in the test added
for the ticket MAILPOET-4365:

https://app.circleci.com/pipelines/github/mailpoet/mailpoet/11041/workflows/8cca6607-2e24-4529-a8ad-cf9a0c663180/jobs/191334/tests#failed-test-0

See MAILPOET-4580 for more details.

[MAILPOET-4365]

lysyjan commented on mailpoet/lib/Entities/SubscriberEntity.php:
```diff
@@ -447,6 +447,12 @@ public function getSegments() {
     });
   }
 
+  public function getSubscribedSegments() {
```
I have a few notes about this new getter.
- There is no annotation with return type
- The name looks very similar to another getter `getSubscriberSegments`
- The name of this getter evokes in me that it returns an array of the SegmentEntity.
- Maybe adding a parameter with status would be better because it allows better reusability.

What do you think about it?

rodrigoprimo replied:
> The name looks very similar to another getter getSubscriberSegments

Yeah, I thought about this but couldn't think of a better name and went it with. But I agree that it has the potential to induce mistakes.

> Maybe adding a parameter with status would be better because it allows better reusability.

What if we add a new optional `$status` parameter to `getSubscriberSegments()`? This way we don't even need to create `getSubscribedSegments()`?

lysyjan replied:
>What if we add a new optional $status parameter to getSubscriberSegments()? This way we don't even need to create getSubscribedSegments()?

This sounds great to me! 👍🏻

rodrigoprimo replied:
@lysyjan, I implemented the approach that we discussed above in https://github.com/mailpoet/mailpoet/pull/4325/commits/422b422d451294e58814528cd6c2b0bf816dcd35. Could you please check this PR again?

lysyjan commented on mailpoet/lib/WooCommerce/Subscription.php:
```diff
@@ -229,60 +237,64 @@ public function handleSubscriberOptin(SubscriberEntity $subscriberEntity, bool $
+    $subscriber->setStatus(SubscriberEntity::STATUS_SUBSCRIBED);
+    if (empty($subscriber->getConfirmedIp()) && empty($subscriber->getConfirmedAt())) {
+      $subscriber->setConfirmedIp(Helpers::getIP());
+      $subscriber->setConfirmedAt(new Carbon());
```
I think that we should use here a time from WP. Something like `Carbon::createFromTimestamp($this->wp->currentTime('timestamp'))`. Do you agree?

rodrigoprimo replied:
Could you please elaborate on why you think we should use current_time() in this case? Is it because in this case we should respect whatever timezone is configured in WP? I haven't thought about it, it would be a change in the current behavior, but not necessarily a problem since what I'm suggesting is already changing the behavior anyway as the original code used MySQL's NOW().

lysyjan replied:
> Could you please elaborate on why you think we should use current_time() in this case?

What I know is that we use `current_time()` when we want to store current time. If I'm not wrong, the reason is to respect the user's timezone. Here is [an example](https://github.com/mailpoet/mailpoet/blob/trunk/mailpoet/lib/Subscription/Pages.php#L190). But I guess it's not critical here.

rodrigoprimo replied:
Yeah, I don't think it is critical and it would change the current behavior which doesn't take into consideration the user timezone set on WP. So I'm inclined to think we don't need to modify it in this PR. Let me know if you agree or not.

lysyjan commented on mailpoet/lib/WooCommerce/Subscription.php:
```diff
@@ -229,60 +237,64 @@ public function handleSubscriberOptin(SubscriberEntity $subscriberEntity, bool $
+    $segmentsIds = (array)$this->settings->get(self::OPTIN_SEGMENTS_SETTING_NAME, []);
+
+    foreach ($segmentsIds as $segmentId) {
+      $segments[] = $this->entityManager->getReference(SegmentEntity::class, $segmentId);
```
Why did you use `getReference` here? This looks a little bit unsafe to me. If I am not wrong, `getRefernece` doesn't load the entity from the database, and it's possible to save invalid data with this approach. For example, when a segment in settings is deleted.
When you use `$this->segmentsRepository->findBy(['id' => $segmentIds])`, the code will be simpler and also it's possible to remove `EntityManager` from dependecies.

rodrigoprimo replied:
I opted to use `getReference` to avoid hitting the database unnecessarily as we don't need the full entity here. But I missed that it was possible for users to delete segments that are added to the `woocommerce.segments` settings and that this would cause problems as you pointed out. I added a new commit using `findBy()` instead of `getReferece()` and simplifying the code.

rodrigoprimo replied:
By the way, do you think we should create a ticket to add some protection against the user deleting segments that are part of `woocommerce.segments`?

lysyjan replied:
>By the way, do you think we should create a ticket to add some protection against the user deleting segments that are part of woocommerce.segments?

👍🏻 I think that we can create a behavior that at least notifies a user about it.

rodrigoprimo replied:
https://mailpoet.atlassian.net/browse/MAILPOET-4597

lysyjan commented on mailpoet/lib/WooCommerce/Subscription.php:
```diff
@@ -169,12 +169,14 @@ public function isCurrentUserSubscribed() {
-    return $subscriberSegment instanceof SubscriberSegment
-      && $subscriberSegment->status === SubscriberEntity::STATUS_SUBSCRIBED;
+    $subscriberSegment = $this->subscriberSegmentRepository->findOneBy(
+      ['subscriber' => $subscriber->getId(), 'segment' => $wcSegment->getId()]
```
I'm not sure, but maybe you don't know it. Passing IDs here is not necessary because the whole entity is enough, but you don't have to change.

rodrigoprimo replied:
I wasn't aware of that. Thanks for mentioning it.

rodrigoprimo committed:
Use findBy() instead of getReference() to get segment entities

Initially, I had opted to use getReference() to avoid querying the
database to get the segment entities as all we need is the segment ID.
But I hadn't realized that this could cause problems if a segment that
is saved in the settings in the option woocommerce.segments is deleted.
Using findBy() here protects against this problem as, if the segment
doesn't exist anymore, it won't be returned.

[MAILPOET-4365]

rodrigoprimo commented:
Thanks for the review, @lysyjan. I replied to your comments. Could you please take a look when you have a moment?

rodrigoprimo committed:
Replace getSubscribedSegments() with a new getSubscriberSegments() parameter

After the PR review, we decided to add a new $status parameter to
getSubscriberSegments() instead of creating getSubscribedSegments().

[MAILPOET-4365]

lysyjan reviewed (approved):
LGTM 👍🏻

Aschepikov reviewed (approved):
This is verified, good job.
